File: ruby/lib/k256/base58.rb
# frozen_string_literal: true

module K256
  # Base58 encoding/decoding utilities for Solana addresses.
  module Base58
    ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    BASE = ALPHABET.length
    INDEXES = ALPHABET.each_char.with_index.to_h.freeze

    class << self
      # Encode bytes to base58 string.
      #
      # @param data [String, Array<Integer>] Bytes to encode
      # @return [String] Base58-encoded string
      def encode(data)
        data = data.bytes if data.is_a?(String)
        return "" if data.empty?

        # Count leading zeros
        leading_zeros = data.take_while(&:zero?).length

        # Convert to big integer
        num = data.reduce(0) { |acc, byte| (acc << 8) + byte }

        # Convert to base58
        result = []
        while num.positive?
          num, remainder = num.divmod(BASE)
          result.unshift(ALPHABET[remainder])
        end

        # Add leading '1's for leading zeros
        ("1" * leading_zeros) + result.join
      end

      # Decode base58 string to bytes.
      #
      # @param str [String] Base58-encoded string
      # @return [Array<Integer>] Decoded bytes
      # @raise [ArgumentError] if string contains invalid characters
      def decode(str)
        return [] if str.empty?

        # Count leading '1's
        leading_ones = str.chars.take_while { |c| c == "1" }.length

        # Convert from base58 to integer
        num = 0
        str.each_char do |c|
          index = INDEXES[c]
          raise ArgumentError, "Invalid Base58 character: #{c}" if index.nil?

          num = num * BASE + index
        end

        # Convert to bytes
        result = []
        while num.positive?
          result.unshift(num & 0xFF)
          num >>= 8
        end

        # Add leading zero bytes
        ([0] * leading_ones) + result
      end

      # Check if a string is a valid Solana public key.
      #
      # @param address [String] Base58-encoded address
      # @return [Boolean] True if valid
      def valid_pubkey?(address)
        return false if address.nil? || address.length < 32 || address.length > 44

        begin
          decoded = decode(address)
          decoded.length == 32
        rescue ArgumentError
          false
        end
      end
    end
  end
end

File: php/src/Utils/Base58.php
<?php

declare(strict_types=1);

namespace K256\Utils;

/**
 * Base58 encoding/decoding utilities for Solana addresses.
 */
final class Base58
{
    private const ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
    private const BASE = 58;

    private static ?array $indexes = null;

    private function __construct()
    {
        // Prevent instantiation
    }

    /**
     * Encode bytes to base58 string.
     *
     * @param string $data Binary data to encode
     * @return string Base58-encoded string
     */
    public static function encode(string $data): string
    {
        if ($data === '') {
            return '';
        }

        $bytes = array_values(unpack('C*', $data));

        // Count leading zeros
        $leadingZeros = 0;
        foreach ($bytes as $byte) {
            if ($byte === 0) {
                $leadingZeros++;
            } else {
                break;
            }
        }

        // Convert to GMP number
        $num = gmp_init(0);
        foreach ($bytes as $byte) {
            $num = gmp_add(gmp_mul($num, 256), $byte);
        }

        // Convert to base58
        $result = '';
        while (gmp_cmp($num, 0) > 0) {
            [$num, $remainder] = gmp_div_qr($num, self::BASE);
            $result = self::ALPHABET[gmp_intval($remainder)] . $result;
        }

        // Add leading '1's
        return str_repeat('1', $leadingZeros) . $result;
    }

    /**
     * Decode base58 string to bytes.
     *
     * @param string $input Base58-encoded string
     * @return string Decoded binary data
     * @throws \InvalidArgumentException if string contains invalid characters
     */
    public static function decode(string $input): string
    {
        if ($input === '') {
            return '';
        }

        self::initIndexes();

        // Count leading '1's
        $leadingOnes = 0;
        for ($i = 0; $i < strlen($input) && $input[$i] === '1'; $i++) {
            $leadingOnes++;
        }

        // Convert from base58 to GMP number
        $num = gmp_init(0);
        for ($i = 0; $i < strlen($input); $i++) {
            $char = $input[$i];
            $index = self::$indexes[$char] ?? -1;
            if ($index < 0) {
                throw new \InvalidArgumentException("Invalid Base58 character: {$char}");
            }
            $num = gmp_add(gmp_mul($num, self::BASE), $index);
        }

        // Convert to bytes
        $bytes = [];
        while (gmp_cmp($num, 0) > 0) {
            [$num, $remainder] = gmp_div_qr($num, 256);
            array_unshift($bytes, gmp_intval($remainder));
        }

        // Add leading zero bytes
        $bytes = array_merge(array_fill(0, $leadingOnes, 0), $bytes);

        return pack('C*', ...$bytes);
    }

    /**
     * Check if a string is a valid Solana public key.
     *
     * @param string $address Base58-encoded address
     * @return bool True if valid
     */
    public static function isValidPubkey(string $address): bool
    {
        if (strlen($address) < 32 || strlen($address) > 44) {
            return false;
        }

        try {
            $decoded = self::decode($address);
            return strlen($decoded) === 32;
        } catch (\InvalidArgumentException) {
            return false;
        }
    }

    private static function initIndexes(): void
    {
        if (self::$indexes !== null) {
            return;
        }

        self::$indexes = [];
        for ($i = 0; $i < strlen(self::ALPHABET); $i++) {
            self::$indexes[self::ALPHABET[$i]] = $i;
        }
    }
}
